<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Analyzing Pathways from PTMs: A Guide • PTMsToPathways</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="96x96" href="../favicon-96x96.png">
<link rel="icon" type="”image/svg+xml”" href="../favicon.svg">
<link rel="apple-touch-icon" sizes="180x180" href="../apple-touch-icon.png">
<link rel="icon" sizes="any" href="../favicon.ico">
<link rel="manifest" href="../site.webmanifest">
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Analyzing Pathways from PTMs: A Guide">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">PTMsToPathways</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.0.0.9000</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../articles/PTMsToPathways.html">Get started</a></li>
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="../articles/RawDataProcessing.html">Analyzing Pathways from PTMs: A Guide</a></li>
  </ul>
</li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Analyzing Pathways from PTMs: A Guide</h1>
                        <h4 data-toc-skip class="author">Nagashree
Avabhrath, Mikhail Ukrainetz, Madison Moffett, Grant Smith, Lucia
Williams, Mark Grimes</h4>
            
      

      <div class="d-none name"><code>RawDataProcessing.Rmd</code></div>
    </div>

    
    
<p>
 
</p>
<div class="section level2">
<h2 id="purpose">Purpose<a class="anchor" aria-label="anchor" href="#purpose"></a>
</h2>
<p>Mass spectrometry data output will vary depending on the experimental
design, source of data, and software used to process the raw spectra. We
start with data in tab-delimited spreadsheet format. (R has a variation
of read.table() called read.csv(), which can be used for comma-delimited
.csv files.) This vignette intends to help users produce a matrix with
PTM names as row names (e.g. FYN p Y411) and numeric data in all the
columns.</p>
</div>
<div class="section level2">
<h2 id="conventions">Conventions<a class="anchor" aria-label="anchor" href="#conventions"></a>
</h2>
<p>We use the following shorthand conventions when describing PTMs. This
can be modified if different modification types are under
investigation.</p>
<p><em>Gene.Name</em> = The HUGO Gene Name is used to identify the
protein/gene</p>
<p><em>Phosphorylation</em> = “p”</p>
<p><em>Lysine acetylation</em> = “ack”</p>
<p><em>Lysine methylation</em> = “kme”</p>
<p><em>Arginine methylation</em> = “rme”</p>
<p><em>Ubiquitination</em> = “ubi”</p>
</div>
<div class="section level2">
<h2 id="data-input">Data input<a class="anchor" aria-label="anchor" href="#data-input"></a>
</h2>
<p>Investigators will need to identify which columns contain key
information for analysis of PTMs. The example raw data file for this
vignette contains only phosphorylation sites (ex_phosphotable =
“phospho_cleaned_mapped.txt”).</p>
<p>“Amino Acid” has the modified amino acid, e.g. S,T, etc.</p>
<p>“Positions Within Proteins” is the amino acid number in the protein
sequence.</p>
<p>Ambiguous modification sites (a modification site whose peptide
sequence is the same in more than one protein) have multiple possible
postions separated by “;”.</p>
<p>“Modification Type” has Phosphorylation” etc.</p>
<p>Note that R converts spaces in column names to periods. After reading
in our data table, the relevant column names are:</p>
<p><strong>genes</strong> = “AllGeneSymbols”</p>
<p><strong>positions</strong> = “Positions.Within.Proteins”</p>
<p><strong>aa</strong> = “Amino.Acid”</p>
<p><strong>modification</strong> = “Modification.Type”</p>
<p>We provide a helper function <em>name.peptide()</em> to make a PTM
name using these columns. This function concatenates ambiguous
modification sites and ignores NA values that are present in cells or
ambiguous gene names.</p>
<p>Many investigators will examine their data in Microsoft Excel, which
can export the spreadsheet in tab- or comma-delimited format.
Unfortunately, Excel inescapably turns some gene names into dates when
present in a cell by themselves. We reverse this with a helper function,
<em>fix.excel()</em>.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Helper functions</span></span>
<span><span class="co"># Function to change dates back into gene names - Excel changes many genes into dates and this can't be turned off!</span></span>
<span><span class="va">fix.excel</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">cell</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="va">fixgenes</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"CDC2"</span>, <span class="st">"1-Sep"</span>, <span class="st">"2-Sep"</span>, <span class="st">"3-Sep"</span>, <span class="st">"4-Sep"</span>, <span class="st">"5-Sep"</span>, <span class="st">"7-Sep"</span>, <span class="st">"8-Sep"</span>, <span class="st">"9-Sep"</span>, <span class="st">"10-Sep"</span>, <span class="st">"11-Sep"</span>, <span class="st">"15-Sep"</span>, <span class="st">"6-Sep"</span>, <span class="st">"1-Oct"</span>, <span class="st">"2-Oct"</span>, <span class="st">"3-Oct"</span>, <span class="st">"4-Oct"</span>, <span class="st">"6-Oct"</span>, <span class="st">"7-Oct"</span>, <span class="st">"11-Oct"</span>, <span class="st">"1-Mar"</span>, <span class="st">"2-Mar"</span>, <span class="st">"3-Mar"</span>, <span class="st">"4-Mar"</span>, <span class="st">"5-Mar"</span>, <span class="st">"6-Mar"</span>, <span class="st">"7-Mar"</span>, <span class="st">"8-Mar"</span>, <span class="st">"9-Mar"</span>, <span class="st">"10-Mar"</span>, <span class="st">"11-Mar"</span>, <span class="st">"C11orf58"</span>, <span class="st">'C17orf57'</span>, <span class="st">'C3orf10'</span>,  <span class="st">'C7orf51'</span>, <span class="st">"C11orf59"</span>, <span class="st">"C4orf16"</span>, <span class="st">"1-Dec"</span>, <span class="st">"14-Sep"</span><span class="op">)</span></span>
<span>  <span class="va">corrects</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"CDK1"</span>, <span class="st">"SEPT1"</span>, <span class="st">"SEPT2"</span>, <span class="st">"SEPT3"</span>, <span class="st">"SEPT4"</span>, <span class="st">"SEPT5"</span>, <span class="st">"SEPT7"</span>, <span class="st">"SEPT8"</span>, <span class="st">"SEPT9"</span>, <span class="st">"SEPT10"</span>, <span class="st">"SEPT11"</span>, <span class="st">"SEPT15"</span>, <span class="st">"SEPT6"</span>, <span class="st">"POU2F1"</span>, <span class="st">"POU2F2"</span>, <span class="st">"POU5F1"</span>, <span class="st">"POU5F1"</span>, <span class="st">"POU3F1"</span>, <span class="st">"POU3F2"</span>, <span class="st">"POU2F3"</span>, <span class="st">"MARCH1"</span>, <span class="st">"MARCH2"</span>, <span class="st">"MARCH3"</span>, <span class="st">"MARCH4"</span>, <span class="st">"MARCH5"</span>, <span class="st">"MARCH6"</span>, <span class="st">"MARCH7"</span>, <span class="st">"MARCH8"</span>, <span class="st">"MARCH9"</span>, <span class="st">"MARCH10"</span>, <span class="st">"MARCH11"</span>, <span class="st">"SMAP"</span>, <span class="st">"EFCAB13"</span>, <span class="st">"BRK1"</span>, <span class="st">"NYAP1"</span>, <span class="st">"LAMTOR1"</span>, <span class="st">'AP1AR'</span>, <span class="st">"DEC1"</span>, <span class="st">"SEPT14"</span><span class="op">)</span></span>
<span>  <span class="va">cellv</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/unlist.html" class="external-link">unlist</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/strsplit.html" class="external-link">strsplit</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/character.html" class="external-link">as.character</a></span><span class="op">(</span><span class="va">cell</span><span class="op">)</span>, <span class="st">"; "</span><span class="op">)</span><span class="op">)</span></span>
<span>  <span class="kw">if</span> <span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/any.html" class="external-link">any</a></span><span class="op">(</span><span class="va">fixgenes</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="va">cellv</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="va">cellv.new</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">gsub</a></span><span class="op">(</span><span class="va">fixgenes</span><span class="op">[</span><span class="va">fixgenes</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="va">cellv</span><span class="op">]</span>, <span class="va">corrects</span><span class="op">[</span><span class="va">fixgenes</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="va">cellv</span><span class="op">]</span>, <span class="va">cellv</span><span class="op">)</span></span>
<span>    <span class="kw"><a href="https://rdrr.io/r/base/function.html" class="external-link">return</a></span> <span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="va">cellv.new</span>, collapse<span class="op">=</span><span class="st">"; "</span><span class="op">)</span><span class="op">)</span></span>
<span>  <span class="op">}</span> <span class="kw">else</span> <span class="kw"><a href="https://rdrr.io/r/base/function.html" class="external-link">return</a></span><span class="op">(</span><span class="va">cell</span><span class="op">)</span>    <span class="op">}</span></span>
<span><span class="co"># This function will handle ambiguous modification sites (a modification site whose peptide sequence is the same in more than one protein) separated by ";" or another separator</span></span>
<span><span class="co"># Make peptide names using this function:</span></span>
<span><span class="va">name.peptide</span> <span class="op">&lt;-</span> <span class="kw">function</span> <span class="op">(</span><span class="va">genes</span>, <span class="va">modification</span><span class="op">=</span><span class="st">"p"</span>, <span class="va">sites</span>, <span class="va">aa</span>, <span class="va">pepsep</span><span class="op">=</span><span class="st">";"</span><span class="op">)</span>   <span class="op">{</span></span>
<span>  <span class="va">genes.v</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/unlist.html" class="external-link">unlist</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/strsplit.html" class="external-link">strsplit</a></span><span class="op">(</span><span class="va">genes</span>, <span class="va">pepsep</span>, fixed <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span></span>
<span>  <span class="va">genes.v</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/which.html" class="external-link">which</a></span><span class="op">(</span><span class="va">genes.v</span> <span class="op">==</span> <span class="st">"NA"</span><span class="op">)</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="cn">NA</span></span>
<span>  <span class="va">genes.v</span> <span class="op">&lt;-</span> <span class="va">genes.v</span><span class="op">[</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="va">genes.v</span><span class="op">)</span><span class="op">]</span></span>
<span>  <span class="va">sites.v</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/unlist.html" class="external-link">unlist</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/strsplit.html" class="external-link">strsplit</a></span><span class="op">(</span><span class="va">sites</span>, <span class="va">pepsep</span>, fixed <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span></span>
<span>  <span class="va">sites.v</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/which.html" class="external-link">which</a></span><span class="op">(</span><span class="va">sites.v</span> <span class="op">==</span> <span class="st">"NA"</span><span class="op">)</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="cn">NA</span></span>
<span>  <span class="va">sites.v</span> <span class="op">&lt;-</span> <span class="va">sites.v</span><span class="op">[</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="va">sites.v</span><span class="op">)</span><span class="op">]</span></span>
<span>  <span class="va">sites.v</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">sapply</a></span><span class="op">(</span><span class="va">sites.v</span>, <span class="kw">function</span> <span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste</a></span> <span class="op">(</span><span class="va">aa</span>, <span class="va">x</span>, sep<span class="op">=</span><span class="st">""</span><span class="op">)</span><span class="op">)</span></span>
<span>  <span class="va">Peptide.v</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/character.html" class="external-link">as.character</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/noquote.html" class="external-link">noquote</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="va">genes.v</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">genes.v</span><span class="op">)</span><span class="op">]</span>, <span class="va">modification</span>, <span class="va">sites.v</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">sites.v</span><span class="op">)</span><span class="op">]</span>, sep<span class="op">=</span><span class="st">" "</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span>  <span class="va">Peptide</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">Peptide.v</span><span class="op">)</span>, collapse<span class="op">=</span><span class="st">"; "</span><span class="op">)</span></span>
<span>  <span class="kw"><a href="https://rdrr.io/r/base/function.html" class="external-link">return</a></span><span class="op">(</span><span class="va">Peptide</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="co"># _________________________________________________________________________________</span></span>
<span><span class="co"># Read in data file that is output from mass spectrometry. </span></span>
<span></span>
<span><span class="co"># You should replace file_path with your own path to file, as in the commented line below</span></span>
<span><span class="va">file_path</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"phospho_cleaned_mapped.txt"</span>, package <span class="op">=</span> <span class="st">"PTMsToPathways"</span><span class="op">)</span></span>
<span><span class="co"># file_path &lt;- "path/to/your/file.txt"</span></span>
<span><span class="va">newphos</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.table</a></span><span class="op">(</span><span class="va">file_path</span>, sep <span class="op">=</span> <span class="st">"\t"</span>, skip <span class="op">=</span> <span class="fl">0</span>, header<span class="op">=</span><span class="cn">TRUE</span>, blank.lines.skip<span class="op">=</span><span class="cn">T</span>, fill<span class="op">=</span><span class="cn">T</span>, quote<span class="op">=</span><span class="st">"\""</span>, dec<span class="op">=</span><span class="st">"."</span>, comment.char <span class="op">=</span> <span class="st">""</span>, stringsAsFactors<span class="op">=</span><span class="cn">F</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># First remove internal control rows (reverse sequences)</span></span>
<span><span class="va">newphos</span> <span class="op">&lt;-</span> <span class="va">newphos</span><span class="op">[</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="va">newphos</span><span class="op">$</span><span class="va">AllGeneSymbols</span><span class="op">)</span>,<span class="op">]</span></span>
<span></span>
<span><span class="co"># If there are dates in the AllGeneSymbols column, use:</span></span>
<span><span class="va">newphos</span><span class="op">$</span><span class="va">AllGeneSymbols</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">sapply</a></span><span class="op">(</span><span class="va">newphos</span><span class="op">$</span><span class="va">AllGeneSymbols</span>, <span class="va">fix.excel</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Next, identify the columns that will be used to name the PTM</span></span>
<span></span>
<span><span class="va">headercols</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"AllGeneSymbols"</span>, <span class="st">"Amino.Acid"</span>, <span class="st">"Positions.Within.Proteins"</span>, <span class="st">"Modification.Type"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Make a separate data frame using these columns. </span></span>
<span><span class="va">newphos.head</span> <span class="op">&lt;-</span> <span class="va">newphos</span><span class="op">[</span>,<span class="va">headercols</span><span class="op">]</span></span>
<span></span>
<span><span class="co"># Now create a unique PTM name (peptide name)</span></span>
<span><span class="va">newphos.head</span><span class="op">$</span><span class="va">Peptide.Name</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/mapply.html" class="external-link">mapply</a></span><span class="op">(</span><span class="va">name.peptide</span>, genes<span class="op">=</span><span class="va">newphos.head</span><span class="op">$</span><span class="va">AllGeneSymbols</span>, sites<span class="op">=</span> <span class="va">newphos.head</span><span class="op">$</span><span class="va">Positions.Within.Proteins</span>, aa<span class="op">=</span><span class="va">newphos.head</span><span class="op">$</span><span class="va">Amino.Acid</span><span class="op">)</span></span></code></pre></div>
<p>Next we identify the data columns, which contain the string
“Intensity.” The example data file is from a multi-PTM study and the
data in this table are from just the phosphorylation pulldown (other
tables are for other PTM types). The optimal pulldown columns are
straightforward to identify by the pulldown strings present in the
sample names, “pTyr” in this case. They are also identifiable by zooming
out and looking at the patterns of missing data, the optimal pulldowns,
as a group, have the least missing data. The second step is to select
colums that have “pTyr.”</p>
<p>These data have technical replicates, which means that the same
samples were run twice. Due to the stochastic selection of peptides for
detection, the pattern of missing values is slightly different between
technical replicates. We therefore merge the technical replicates taking
the value of either replicate where it’s missing in the other, and
averaging values detected in both, using the function,
<em>merge2cols()</em>.</p>
</div>
<div class="section level2">
<h2 id="key-for-column-names">Key for column names<a class="anchor" aria-label="anchor" href="#key-for-column-names"></a>
</h2>
<p><strong>C</strong> = Crizotinib</p>
<p><strong>D</strong> = DMSO</p>
<p><strong>E</strong> = Erlotinib</p>
<p><strong>Pr</strong> = PR171</p>
<p><strong><em>For Intensity columns</em></strong></p>
<p><em>C1-1</em>: Crizotinib biological replicate 1- technical replicate
1</p>
<p><em>C1-2</em>: Crizotinib biological replicate 1- technical replicate
2</p>
<p><em>C2-1</em>: Crizotinib biological replicate 2- technical replicate
1</p>
<p><em>C2-2</em>: Crizotinib biological replicate 2- technical replicate
2</p>
<p><em>C3-1</em>: Crizotinib biological replicate 3- technical replicate
1</p>
<p><em>C3-2</em>: Crizotinib biological replicate 3- technical replicate
2</p>
<p>… &amp; similar for other drugs</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Define Data columns</span></span>
<span><span class="va">phosdata</span> <span class="op">&lt;-</span> <span class="va">newphos</span><span class="op">[</span>,<span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">grep</a></span><span class="op">(</span><span class="st">"Intensity"</span>, <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">newphos</span><span class="op">)</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="va">phosdata</span> <span class="op">&lt;-</span> <span class="va">phosdata</span><span class="op">[</span>,<span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">grep</a></span><span class="op">(</span><span class="st">"pTyr"</span>, <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">phosdata</span><span class="op">)</span><span class="op">)</span><span class="op">]</span></span>
<span></span>
<span><span class="co"># Simplify column names</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">phosdata</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">sapply</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">phosdata</span><span class="op">)</span>, <span class="kw">function</span> <span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="fu"><a href="https://rdrr.io/r/base/unlist.html" class="external-link">unlist</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/strsplit.html" class="external-link">strsplit</a></span><span class="op">(</span><span class="va">x</span>, <span class="st">"Intensity."</span><span class="op">)</span><span class="op">)</span><span class="op">[</span><span class="fl">2</span><span class="op">]</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Make zero into NA, which it is. (Note that this may not apply if you are confident that zero means actual zero, which is possible with certain technical advances.) </span></span>
<span><span class="va">zer0</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/which.html" class="external-link">which</a></span><span class="op">(</span><span class="va">phosdata</span><span class="op">==</span><span class="fl">0</span>, arr.ind <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="va">phosdata</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/replace.html" class="external-link">replace</a></span> <span class="op">(</span><span class="va">phosdata</span>, <span class="va">zer0</span>, <span class="cn">NA</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Define technical replicates</span></span>
<span><span class="va">tr1.opt</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">phosdata</span><span class="op">)</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">grep</a></span><span class="op">(</span><span class="st">".1"</span>, <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">phosdata</span><span class="op">)</span>, fixed<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="va">tr2.opt</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">phosdata</span><span class="op">)</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">grep</a></span><span class="op">(</span><span class="st">".2"</span>, <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">phosdata</span><span class="op">)</span>, fixed<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span><span class="op">]</span></span>
<span></span>
<span><span class="co"># Use this function to average technical replicates. This function ignores NA values in either column and takes the average in the case where there are two values.</span></span>
<span><span class="va">merge2cols</span> <span class="op">&lt;-</span> <span class="kw">function</span> <span class="op">(</span><span class="va">colv1</span>, <span class="va">colv2</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="va">newcolv</span><span class="op">=</span><span class="cn">NA</span></span>
<span>  <span class="kw">if</span> <span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="va">colv1</span><span class="op">)</span> <span class="op">&amp;</span> <span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="va">colv2</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="va">newcolv</span><span class="op">=</span><span class="cn">NA</span></span>
<span>    <span class="kw"><a href="https://rdrr.io/r/base/function.html" class="external-link">return</a></span><span class="op">(</span><span class="va">newcolv</span><span class="op">)</span><span class="op">}</span> <span class="kw">else</span></span>
<span>      <span class="kw">if</span> <span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="va">colv1</span><span class="op">)</span> <span class="op">|</span> <span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="va">colv2</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>        <span class="va">newcolv</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span><span class="va">colv1</span>, <span class="va">colv2</span>, na.rm<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span>
<span>        <span class="kw"><a href="https://rdrr.io/r/base/function.html" class="external-link">return</a></span><span class="op">(</span><span class="va">newcolv</span><span class="op">)</span> <span class="op">}</span> <span class="kw">else</span></span>
<span>          <span class="va">newcolv</span> <span class="op">&lt;-</span> <span class="op">(</span><span class="va">colv1</span> <span class="op">+</span> <span class="va">colv2</span><span class="op">)</span><span class="op">/</span><span class="fl">2</span></span>
<span>        <span class="kw"><a href="https://rdrr.io/r/base/function.html" class="external-link">return</a></span><span class="op">(</span><span class="va">newcolv</span><span class="op">)</span> <span class="op">}</span></span>
<span></span>
<span><span class="co"># Use merge2cols() function to average technical replciates:</span></span>
<span><span class="va">phosdata.merged</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">matrix</a></span><span class="op">(</span>nrow<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">phosdata</span><span class="op">)</span>, ncol<span class="op">=</span><span class="fl">18</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="kw">for</span><span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">tr1.opt</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="va">phosdata.merged</span><span class="op">[</span>,<span class="va">i</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/mapply.html" class="external-link">mapply</a></span><span class="op">(</span><span class="va">merge2cols</span>, colv1<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/numeric.html" class="external-link">as.numeric</a></span><span class="op">(</span><span class="va">phosdata</span><span class="op">[</span>, <span class="va">tr1.opt</span><span class="op">[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span><span class="op">)</span>, colv2<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/numeric.html" class="external-link">as.numeric</a></span><span class="op">(</span><span class="va">phosdata</span><span class="op">[</span>,<span class="va">tr2.opt</span><span class="op">[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">phosdata.merged</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">sapply</a></span><span class="op">(</span><span class="va">tr1.opt</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="fu"><a href="https://rdrr.io/r/base/substr.html" class="external-link">substr</a></span><span class="op">(</span><span class="va">x</span>, start<span class="op">=</span><span class="fl">1</span>, stop<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/nchar.html" class="external-link">nchar</a></span><span class="op">(</span><span class="va">x</span><span class="op">)</span><span class="op">-</span><span class="fl">2</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Merge with header</span></span>
<span><span class="va">phosdatafile</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/cbind.html" class="external-link">cbind</a></span><span class="op">(</span><span class="va">newphos.head</span>, <span class="va">phosdata.merged</span><span class="op">)</span></span>
<span><span class="co"># This file could be safed for reference using write.table()</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/write.table.html" class="external-link">write.table</a></span><span class="op">(</span><span class="va">phosdatafile</span>, file<span class="op">=</span><span class="st">"phosdatafile.txt"</span>, row.names <span class="op">=</span> <span class="cn">FALSE</span>, sep<span class="op">=</span><span class="st">"\t"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># For subsequent steps</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">phosdatafile</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">phosdatafile</span><span class="op">$</span><span class="va">Peptide.Name</span></span>
<span><span class="va">phosdata.df</span> <span class="op">&lt;-</span> <span class="va">phosdatafile</span><span class="op">[</span>,<span class="fl">6</span><span class="op">:</span><span class="fl">23</span><span class="op">]</span></span>
<span><span class="co"># This contains just the data with individual PTMs as rownames</span></span>
<span></span>
<span><span class="co"># Log base 2 transformation improves clustering.</span></span>
<span><span class="va">log2phosdata</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log2</a></span><span class="op">(</span><span class="va">phosdata.df</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># This could be used as the ptmtable for subsequent clustering.</span></span>
<span><span class="co"># ptmtable &lt;- log2phosdata</span></span></code></pre></div>
<p>For experiments where treatment with drugs is compared to control
samples, adding treatment/control ratios as additional data column can
improve clustering. This optional step adds dimensions to the data set
that enhance focus on the changes in response to drug treatments.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="co"># Simplify column names first:</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">phosdata.df</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">sapply</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">phosdata.df</span><span class="op">)</span>, <span class="kw">function</span> <span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/unlist.html" class="external-link">unlist</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/strsplit.html" class="external-link">strsplit</a></span><span class="op">(</span><span class="va">x</span>, <span class="st">"SEPTM_pTyr"</span><span class="op">)</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>, <span class="fu"><a href="https://rdrr.io/r/base/unlist.html" class="external-link">unlist</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/strsplit.html" class="external-link">strsplit</a></span><span class="op">(</span><span class="va">x</span>, <span class="st">"SEPTM_pTyr"</span><span class="op">)</span><span class="op">)</span><span class="op">[</span><span class="fl">2</span><span class="op">]</span>, sep<span class="op">=</span><span class="st">""</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Explore using ratios where control=rowMeans (D1, D2, D3)</span></span>
<span></span>
<span><span class="va">H3122control</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colSums.html" class="external-link">rowMeans</a></span><span class="op">(</span><span class="va">phosdata.df</span><span class="op">[</span>, <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">phosdata.df</span><span class="op">)</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">grep</a></span><span class="op">(</span><span class="st">"H3122.D"</span>, <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">phosdata.df</span><span class="op">)</span><span class="op">)</span><span class="op">]</span><span class="op">]</span>, na.rm<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="co"># Change NaN to NA</span></span>
<span><span class="va">H3122control</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/is.finite.html" class="external-link">is.nan</a></span><span class="op">(</span><span class="va">H3122control</span><span class="op">)</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="cn">NA</span></span>
<span></span>
<span><span class="va">PC9control</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colSums.html" class="external-link">rowMeans</a></span><span class="op">(</span><span class="va">phosdata.df</span><span class="op">[</span>, <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">phosdata.df</span><span class="op">)</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">grep</a></span><span class="op">(</span><span class="st">"PC9.D"</span>, <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">phosdata.df</span><span class="op">)</span><span class="op">)</span><span class="op">]</span><span class="op">]</span>, na.rm<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="co"># Change NaN to NA</span></span>
<span><span class="va">PC9control</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/is.finite.html" class="external-link">is.nan</a></span><span class="op">(</span><span class="va">PC9control</span><span class="op">)</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="cn">NA</span></span>
<span></span>
<span><span class="co"># Calculate treatment/control ratios</span></span>
<span><span class="co"># H3122 cells</span></span>
<span><span class="va">H3122.C1.ratio</span> <span class="op">&lt;-</span> <span class="va">phosdata.df</span><span class="op">$</span><span class="va">H3122.C1</span><span class="op">/</span><span class="va">H3122control</span></span>
<span><span class="va">H3122.C2.ratio</span> <span class="op">&lt;-</span> <span class="va">phosdata.df</span><span class="op">$</span><span class="va">H3122.C2</span><span class="op">/</span><span class="va">H3122control</span></span>
<span><span class="va">H3122.C3.ratio</span> <span class="op">&lt;-</span> <span class="va">phosdata.df</span><span class="op">$</span><span class="va">H3122.C3</span><span class="op">/</span><span class="va">H3122control</span></span>
<span><span class="va">H3122.PR1.ratio</span> <span class="op">&lt;-</span> <span class="va">phosdata.df</span><span class="op">$</span><span class="va">H3122.PR1</span><span class="op">/</span><span class="va">H3122control</span></span>
<span><span class="va">H3122.PR2.ratio</span> <span class="op">&lt;-</span> <span class="va">phosdata.df</span><span class="op">$</span><span class="va">H3122.PR2</span><span class="op">/</span><span class="va">H3122control</span></span>
<span><span class="va">H3122.PR3.ratio</span> <span class="op">&lt;-</span> <span class="va">phosdata.df</span><span class="op">$</span><span class="va">H3122.PR3</span><span class="op">/</span><span class="va">H3122control</span></span>
<span><span class="co"># PC9 cells</span></span>
<span><span class="va">PC9.E1.ratio</span> <span class="op">&lt;-</span> <span class="va">phosdata.df</span><span class="op">$</span><span class="va">PC9.E1</span><span class="op">/</span><span class="va">PC9control</span></span>
<span><span class="va">PC9.E2.ratio</span> <span class="op">&lt;-</span> <span class="va">phosdata.df</span><span class="op">$</span><span class="va">PC9.E2</span><span class="op">/</span><span class="va">PC9control</span></span>
<span><span class="va">PC9.E3.ratio</span> <span class="op">&lt;-</span> <span class="va">phosdata.df</span><span class="op">$</span><span class="va">PC9.E3</span><span class="op">/</span><span class="va">PC9control</span></span>
<span><span class="va">PC9.PR1.ratio</span> <span class="op">&lt;-</span> <span class="va">phosdata.df</span><span class="op">$</span><span class="va">PC9.PR1</span><span class="op">/</span><span class="va">PC9control</span></span>
<span><span class="va">PC9.PR2.ratio</span> <span class="op">&lt;-</span> <span class="va">phosdata.df</span><span class="op">$</span><span class="va">PC9.PR2</span><span class="op">/</span><span class="va">PC9control</span></span>
<span><span class="va">PC9.PR3.ratio</span> <span class="op">&lt;-</span> <span class="va">phosdata.df</span><span class="op">$</span><span class="va">PC9.PR3</span><span class="op">/</span><span class="va">PC9control</span></span>
<span></span>
<span><span class="co"># Put these columns in a data frame</span></span>
<span><span class="va">phos_ratios</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span><span class="va">H3122.C1.ratio</span>, <span class="va">H3122.C2.ratio</span>, <span class="va">H3122.C3.ratio</span>, <span class="va">H3122.PR1.ratio</span>, <span class="va">H3122.PR2.ratio</span>, <span class="va">H3122.PR3.ratio</span>, <span class="va">PC9.E1.ratio</span>, <span class="va">PC9.E2.ratio</span>, <span class="va">PC9.E3.ratio</span>, <span class="va">PC9.PR1.ratio</span>, <span class="va">PC9.PR2.ratio</span>, <span class="va">PC9.PR3.ratio</span><span class="op">)</span></span>
<span><span class="co"># Check</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/identical.html" class="external-link">identical</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">phosdata.df</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">phos_ratios</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co"># TRUE</span></span>
<span><span class="co"># Make limits to unweight extreme values. This has been shown to improve clustering, and a ratio of 1000 is biologically not really functional different than a ratio of 100. </span></span>
<span><span class="va">hi.ratio</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/which.html" class="external-link">which</a></span><span class="op">(</span><span class="va">phos_ratios</span><span class="op">&gt;=</span><span class="fl">100</span>, arr.ind <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="va">low.ratio</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/which.html" class="external-link">which</a></span><span class="op">(</span><span class="va">phos_ratios</span> <span class="op">&lt;=</span><span class="fl">1</span><span class="op">/</span><span class="fl">100</span>, arr.ind <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="va">phos_ratios.lim</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/replace.html" class="external-link">replace</a></span> <span class="op">(</span><span class="va">phos_ratios</span>, <span class="va">hi.ratio</span>, <span class="fl">100</span><span class="op">)</span> </span>
<span><span class="va">phos_ratios.lim</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/replace.html" class="external-link">replace</a></span> <span class="op">(</span><span class="va">phos_ratios.lim</span>, <span class="va">low.ratio</span>, <span class="fl">1</span><span class="op">/</span><span class="fl">100</span><span class="op">)</span> </span>
<span><span class="co"># log2 transformation improves clustering </span></span>
<span><span class="va">phos_ratios.lim.log2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log2</a></span><span class="op">(</span><span class="va">phos_ratios.lim</span><span class="op">)</span></span>
<span><span class="co"># lucy comment: this was below but we needed phosdata_plus_ratios</span></span>
<span><span class="va">phosdata_plus_ratios</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/cbind.html" class="external-link">cbind</a></span><span class="op">(</span><span class="va">log2phosdata</span>, <span class="va">phos_ratios.lim.log2</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/boxplot.html" class="external-link">boxplot</a></span><span class="op">(</span><span class="va">phosdata_plus_ratios</span><span class="op">)</span></span></code></pre></div>
<p><img src="plots/Example%20ratio%20calculations-1.png" width="100%" style="display: block; margin: auto;"></p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Now merge intensity and ratio data</span></span>
<span><span class="co"># Check</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/identical.html" class="external-link">identical</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">phos_ratios.lim.log2</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">log2phosdata</span><span class="op">)</span><span class="op">)</span> </span>
<span><span class="co"># TRUE, okay</span></span>
<span></span>
<span></span>
<span><span class="co"># This can be used as the example ptmtable for subsequent testing.</span></span>
<span><span class="co"># ptmtable &lt;- phosdata_plus_ratios</span></span>
<span><span class="co">#&gt; [1] TRUE</span></span>
<span><span class="co">#&gt; [1] TRUE</span></span></code></pre></div>
<p>For experiments involving multiple PTMs, or if investigators wish to
combine several data sets, the data can be combined. For example,
pulldowns were made to isolate acetylated and ubiquitinated peptides
from the same experimental samples. Combining the data is simply
repeating the above steps using the correct optimum data columns, then
making the column names the same, and binding all the rows together.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">######################</span></span>
<span><span class="co"># Workflow: Repeat the with aceltylation or other PTM data (ubiquitination in this case).</span></span>
<span><span class="co"># This can be done with data that include ratios as above if desired.</span></span>
<span><span class="co"># How to put them together:</span></span>
<span><span class="co"># Make column names common. For example:</span></span>
<span><span class="va">kgp</span> <span class="op">&lt;-</span> <span class="va">phosdata.df</span></span>
<span><span class="co"># lucy comment: ackdata.df not defined, so I'm just commenting everything out below</span></span>
<span><span class="co"># kga &lt;- ackdata.df</span></span>
<span><span class="co"># kgu &lt;- ubidata.df</span></span>
<span><span class="co"># names(kgp) &lt;- sapply(names(kgp), function (x) paste(unlist(strsplit(x, "_pTyr"))[1], unlist(strsplit(x, "_pTyr"))[2], sep=""))</span></span>
<span><span class="co"># names(kga) &lt;- sapply(names(kga), function (x) paste(unlist(strsplit(x, "_AcK"))[1], unlist(strsplit(x, "_AcK"))[2], sep=""))</span></span>
<span><span class="co"># names(kgu) &lt;- sapply(names(kgu), function (x) paste(unlist(strsplit(x, "_Ubi"))[1], unlist(strsplit(x, "_Ubi"))[2], sep=""))</span></span>
<span><span class="co"># identical(names(kgp), names(kga)) # Check TRUE</span></span>
<span><span class="co">#</span></span>
<span><span class="co"># ptmdata &lt;- rbind (kgp, kga, kgu) </span></span>
<span><span class="co"># Reorder here</span></span>
<span><span class="co"># ptmdata &lt;- ptmdata[order(rownames(ptmdata)),]</span></span>
<span></span>
<span><span class="co"># This optional step improves clustering in our hands:</span></span>
<span><span class="co"># log2ptmdata &lt;- log2(ptmdata)</span></span>
<span></span>
<span><span class="co"># This could be used as the ptmtable for clustering.</span></span>
<span><span class="co"># ptmtable = log2ptmdata</span></span></code></pre></div>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Mark Grimes, Mikhail Ukrainetz, Grant Smith, Madison Moffett, Nagashree Avabhrath, Lucia Williams.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer>
</div>





  </body>
</html>
